webanno-api-dao/src/main/java/de/tudarmstadt/ukp/clarin/webanno/api/dao/CasStorageServiceImpl.java

```diff
@@ -999,9 +1007,17 @@ public void close()
         {
             if (holder != null) {
                 try {
-                    log.trace("Returning briefly borrowed CAS [{}]@[{}]({})", username,
-                            documentName, documentId);
-                    exclusiveAccessPool.returnObject(key, holder);
+                    if (invalidateOnClose) {
+                        log.trace("Invalidating CAS [{}]@[{}]({})", username,
+                                documentName, documentId);
+                        exclusiveAccessPool.invalidateObject(key, holder);
+                        exclusiveAccessHolders.remove(holder);
```

Comment:
Why do we need to remove the holder? According to the comment in `CasHolder borrowCas(CasKey aKey)`, that should not be necessary:

```
            CasHolder holder = exclusiveAccessPool.borrowObject(aKey);
            // Add the holder to the set of known holder. Because this set it using weak
            // references, and because we use the set only to inform holders when they become
            // invalid we do never have to explicitly remove the holder from the set
            exclusiveAccessHolders.add(holder);
```

Reply:
You are right, it should not be necessary. it made the most sense to me at the time because this way we really delete all cas instances of cases that no longer exist. Otherwise we still have an outdated version in the holder.